backends/avx/ceed-avx-basis.c

```diff
@@ -78,31 +49,247 @@ static int CeedTensorContract8_Avx(Ceed ceed, CeedInt A, CeedInt B,
+  CeedInt tstride0 = B, tstride1 = 1;
+  if (tmode == CEED_TRANSPOSE) {
+    tstride0 = 1; tstride1 = J;
+  }
+
+  CeedInt Jbreak = J%JJ ? (J/JJ)*JJ : (J/JJ-1)*JJ;
+  for (CeedInt a=0; a<A; a++) {
+    // Blocks of 4 columns
+    for (CeedInt c = (C/CC)*CC; c<C; c+=4) {
+      // Blocks of 4 rows
+      for (CeedInt j=0; j<Jbreak; j+=JJ) {
+        __m256d vv[JJ]; // Output tile to be held in registers
+        for (CeedInt jj=0; jj<JJ; jj++)
+          vv[jj] = _mm256_loadu_pd(&v[(a*J+j+jj)*C+c]);
+
+        for (CeedInt b=0; b<B; b++) {
+          __m256d tqu;
+          if (C-c == 1)
+            tqu = _mm256_set_pd(0.0, 0.0, 0.0, u[(a*B+b)*C+c+0]);
+          else if (C-c == 2)
+            tqu = _mm256_set_pd(0.0, 0.0, u[(a*B+b)*C+c+1],
+                                u[(a*B+b)*C+c+0]);
+          else if (C-c == 3)
+            tqu = _mm256_set_pd(0.0, u[(a*B+b)*C+c+2], u[(a*B+b)*C+c+1],
+                                u[(a*B+b)*C+c+0]);
+          else
+            tqu = _mm256_loadu_pd(&u[(a*B+b)*C+c]);
+          for (CeedInt jj=0; jj<JJ; jj++) // unroll
+            vv[jj] += _mm256_mul_pd(tqu,
+                                    _mm256_set1_pd(t[(j+jj)*tstride0 + b*tstride1]));
+        }
+        for (CeedInt jj=0; jj<JJ; jj++)
+          _mm256_storeu_pd(&v[(a*J+j+jj)*C+c], vv[jj]);
       }
     }
-    // Any remainder
-    for (CeedInt j=(J/JJ)*JJ; j<J; j++) {
+    // Remainder of rows, all columns
+    for (CeedInt j=Jbreak; j<J; j++)
       for (CeedInt b=0; b<B; b++) {
         CeedScalar tq = t[j*tstride0 + b*tstride1];
-        for (CeedInt c=0; c<C; c++)
+        for (CeedInt c=(C/CC)*CC; c<C; c++)
           v[(a*J+j)*C+c] += tq * u[(a*B+b)*C+c];
       }
+  }
+  return 0;
+}
+
+// Serial Tensor Contract C=1 Case
+static inline int CeedTensorContract_Avx_Single(Ceed ceed, CeedInt A,
+    CeedInt B, CeedInt C, CeedInt J, const CeedScalar *restrict t,
+    CeedTransposeMode tmode, const CeedInt Add, const CeedScalar *restrict u,
+    CeedScalar *restrict v, const CeedInt AA, const CeedInt JJ) {
+  CeedInt tstride0 = B, tstride1 = 1;
+  if (tmode == CEED_TRANSPOSE) {
+    tstride0 = 1; tstride1 = J;
+  }
+
+  if (!Add)
+    for (CeedInt q=0; q<A*J*C; q++)
+      v[q] = (CeedScalar) 0.0;
+
+  // Blocks of 4 rows
+  for (CeedInt a=0; a<(A/AA)*AA; a+=AA) {
+    for (CeedInt j=0; j<(J/JJ)*JJ; j+=JJ) {
+      __m256d vv[AA][JJ/4]; // Output tile to be held in registers
+      for (CeedInt aa=0; aa<AA; aa++)
+        for (CeedInt jj=0; jj<JJ/4; jj++)
+          vv[aa][jj] = _mm256_loadu_pd(&v[(a+aa)*J+j+jj*4]);
+
+      for (CeedInt b=0; b<B; b++) {
+        for (CeedInt jj=0; jj<JJ/4; jj++) { // unroll
+          __m256d tqv = _mm256_set_pd(t[(j+jj*4+3)*tstride0 + b*tstride1],
+                                      t[(j+jj*4+2)*tstride0 + b*tstride1],
+                                      t[(j+jj*4+1)*tstride0 + b*tstride1],
+                                      t[(j+jj*4+0)*tstride0 + b*tstride1]);
```

Comment:
Just browsing the code, it isn't obvious to me why this is in reverse order. Do I just need more context or would a comment help?